docs/foreign_code.md

```diff
@@ -0,0 +1,156 @@
+
+Sandboxes are difficult to program with in C-like languages because there is no useful construct in the abstract machine that can represent them.
+Web browsers are the most aggressively compartmentalised applications and yet run with a handful of different sandbox types.
+Because there is no abstraction in C for distinct sets of objects with mediated entry points, sandboxing code written in C leaks mechanism into the security code (for example, the idea that sandboxes are process).
+Any call between sandboxes must go via some RPC layer.
+Allocating memory inside a sandbox from outside and directly manipulating data structures is difficult to get right because the language doesn't impose any restrictions on where any given pointer can be stored.
+
+A Verona region is a set of related objects, with arbitrary topology, that have a single entry point.
+Sandboxes map cleanly to this abstraction.
+A sandbox's heap can contain pointers to any other data structures owned by the sandbox.
+It cannot contain pointers outside of the sandbox because following them would violate the sandbox policy.
+The Verona type system enforces this automatically by disallowing cross-region pointers.
+
+We hope that the region abstraction will make programming with sandboxes natural within the language.
+
+### Foreign libraries are classes
+
+If the unit of interoperability is a library, how do we surface that within Verona?
+We aim to treat foreign libraries just like any other module: as classes that can contain nested types.
+For example, we may import LLVM in something like this (straw-man syntax):
+
+```verona
+class LLVM
+{
+	[#native, #interface_language="c++"]
+	imports library("libLLVM.so", "llvm.h");
+}
+```
+
+This includes two attributes that tell the compiler the format of the library (native code) and how to parse the header (which, in this example, would have to be something provided by the user that included the desired headers).
+Once C++ module support is more widespread, module imports can take the place of header inclusions.
+
+This instructs the compiler to take the C++ namespace structure and inject each namespace as a class within the LLVM class and all C++ classes as nested classes.
+
+Instantiating this class would require the user to specify a sandboxing policy, for example (again, straw-man syntax):
+
+```verona
+var llvm = LLVM.Create(PROCESS);
+```
+
+This would create a new process that loaded the LLVM library and had RPC invocations for any of the LLVM functions that this program called.
+The `llvm` object in this example is a new isolated region.
+The programmer can then construct LLVM types within that region.
+
+Implementation
+--------------
+
+The proposed implementation strategy for C/C++ interoperability is to deeply embed clang.
+We are already depending on LLVM (and MLIR) for our compiler infrastructure and so embedding clang is not a significant addition.
```

Comment:
Does this mean interoperation with other languages will also need to embed their front-ends?

Rust and any other LLVM-based language should be ok (module different LLVM versions they require), but we also want to embed C# and other non-LLVM based languages, no?

It would be bad to have two-tiered support for external languages.

Reply:
Not necessarily.  C/C++ are among the most difficult cases for two reasons:

 - Lack of a clear module system or interface format.  C/C++ 'interfaces' have to be extracted by compiling the code and cannot be recreated with 100% fidelity by anything that is not a complete C/C++ preprocess, parser, and semantic analysis toolchain.
 - ABI complexity.  The standard C ABI is not too hard to handle for any given platform (except bitfields) but there are a load of non-standard GCC and MSVC extensions that give fine-grained control over structure layout.  Beyond that, C++ is even harder. Things like diamond inheritance (handled differently in Itanium / MSVC ABIs), RTTI (handled completely differently in MSVC vs Itanium, subtly differently in Itanium vs Aarch32 vs Fuchsia), exception ABIs (totally different between Win32, Win64, and Itanium, with some significant differences between ARM and x86 Windows). Getting these right is really, really hard in anything that isn't a complete C/C++ compiler.

Most other languages that we'd want to interop with have some combination of simpler ABIs, cleaner module / interface definitions, and so on.  For .NET, interfaces are well defined and the binary format for assemblies is specified, as are various bits of the interop layer, so it should be fairly easy to add a .NET interface without needing to embed a .NET compiler.

Reply:
I've added a suggestion to add most of this comment to the text above.